File: app/src/main/java/com/example/knocker/model/MyReceiver.java
package com.example.knocker.model;

import android.annotation.TargetApi;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.telephony.SmsMessage;
import android.util.Log;
import android.widget.Toast;

import java.util.Objects;

public class MyReceiver extends BroadcastReceiver {

    private static final String SMS_RECEIVED = MyReceiver.class.getSimpleName();
    private static final String pdu_type = "pdus";
    private static final String TAG = "SmsBroadcastReceiver";
    private String msg, phoneNumber = "";

    @TargetApi(Build.VERSION_CODES.M)
    @Override
    public void onReceive(Context context, Intent intent) {

        // Get the SMS message.
        Bundle bundle = intent.getExtras();
        SmsMessage[] msgs;
        String strMessage = "";
        assert bundle != null;
        String format = bundle.getString("format");
        // Retrieve the SMS message received.
        Object[] pdus = (Object[]) bundle.get(pdu_type);
        if (pdus != null) {
            // Check the Android version.
            boolean isVersionM =
                    (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M);
            // Fill the msgs array.
            msgs = new SmsMessage[pdus.length];
            for (int i = 0; i < msgs.length; i++) {
                // Check Android version and use appropriate createFromPdu.
                if (isVersionM) {
                    // If Android version M or newer:
                    msgs[i] = SmsMessage.createFromPdu((byte[]) pdus[i], format);
                } else {
                    // If Android version L or older:
                    msgs[i] = SmsMessage.createFromPdu((byte[]) pdus[i]);
                }
                // Build the message to show.
                strMessage += "SMS from " + msgs[i].getOriginatingAddress();
                strMessage += " :" + msgs[i].getMessageBody() + "\n";
                // Log and display the SMS message.
                Log.d(TAG, "onReceive: " + strMessage);
                Toast.makeText(context, strMessage, Toast.LENGTH_LONG).show();

//
//        Log.i(TAG, "Intent Received : " + intent.getAction());
//
//        if (Objects.equals(intent.getAction(), SMS_RECEIVED)) {
//            Bundle dataBundle = intent.getExtras();
//
//            if (dataBundle != null) {
//                Object[] mypdu = (Object[]) dataBundle.get("mypdu");
//                assert mypdu != null;
//                final SmsMessage[] message = new SmsMessage[mypdu.length];
//
//                for (int i = 0; i < mypdu.length; i++) {
//                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
//                        String format = dataBundle.getString("format");
//                        message[i] = SmsMessage.createFromPdu((byte[]) mypdu[i], format);
//                    } else {
//                        message[i] = SmsMessage.createFromPdu((byte[]) mypdu[i]);
//                    }
//                    msg = message[i].getMessageBody();
//                    phoneNumber = message[i].getOriginatingAddress();
//                }
//                Toast.makeText(context, "Message: " + msg + "\n Number: " + phoneNumber, Toast.LENGTH_LONG).show();
            }
        }
    }
}

File: app/src/main/java/com/example/knocker/model/MyReceiver.kt
package com.example.knocker.model

import android.annotation.SuppressLint
import android.annotation.TargetApi
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Build
import android.telephony.SmsMessage
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.ListView
import android.widget.Toast

import com.example.knocker.R
import com.example.knocker.controller.MessageListAdapter

import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Calendar

@Suppress("DEPRECATION")
class MyReceiver : BroadcastReceiver() {

    private val compose_message_listOfMessage = ArrayList<Message>()

    private var composeMessageView: View? = null

    @SuppressLint("InflateParams", "RtlHardcoded")
    @TargetApi(Build.VERSION_CODES.M)
    override fun onReceive(context: Context, intent: Intent) {

        // Get the SMS message.
        val bundle = intent.extras
        val msgs: Array<SmsMessage?>

        assert(bundle != null)
        val format = bundle!!.getString("format")

        // Retrieve the SMS message received.
        val pdus = bundle.get(pdu_type) as Array<*>

        // Check the Android version.
        val isVersionM = Build.VERSION.SDK_INT >= Build.VERSION_CODES.M

        // Fill the msgs array.
        msgs = arrayOfNulls(pdus.size)

        for (i in msgs.indices) {
            // Check Android version and use appropriate createFromPdu.
            if (isVersionM) {
                // If Android version M or newer:
                msgs[i] = SmsMessage.createFromPdu(pdus[i] as ByteArray, format)
            } else {
                // If Android version L or older:
                msgs[i] = SmsMessage.createFromPdu(pdus[i] as ByteArray)
            }

            // Build the message to show.
            val phoneNumber = msgs[i]!!.originatingAddress
            val msg = msgs[i]!!.messageBody + "\n"

            // Log and display the SMS message.
            Log.d(TAG, "onReceive: $msg")

            Toast.makeText(context, msg + " from : " + phoneNumber, Toast.LENGTH_LONG).show();

//            val inflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
//            composeMessageView = inflater.inflate(R.layout.activity_compose_message, null)
//
//            val c = Calendar.getInstance().time
//
//            val date = SimpleDateFormat("dd-MMM-yyyy")
//            val hour = SimpleDateFormat("hh:mm:ss")
//            val currentDate = date.format(c)
//            val currentHour = hour.format(c)
//
//            val messageClass = Message(msg, false, phoneNumber, 0, currentDate, currentHour)
//
//            compose_message_listOfMessage.add(messageClass)
//
//            callComposeMessageLayout(context, composeMessageView)
        }
    }

    private fun callComposeMessageLayout(context: Context, view: View?) {
        val compose_message_ListViewMessage = view!!.findViewById<ListView>(R.id.compose_message_list_view_message)
        compose_message_ListViewMessage.adapter = MessageListAdapter(context, compose_message_listOfMessage)
    }

    companion object {

        private val SMS_RECEIVED = MyReceiver::class.java.simpleName
        private val pdu_type = "pdus"
        private val TAG = "SmsBroadcastReceiver"
    }
}